<!DOCTYPE html>
<html lang="en">

<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <!-- jQuery first, then D3.js, then Bootstrap JS -->
    <script type="text/javascript" src="../static/js/jquery.min.js"></script>
    <script type="text/javascript" src="../static/js/d3.v6.min.js"></script>
    <link rel="stylesheet" href="../static/css/common.css">
    <script type="text/javascript" src="../static/js/haplotype-plot.js"></script>
    <script type="text/javascript" src="../static/js/utils.js"></script>
    <title>Gene Haplotypes</title>
</head>

<body>
    <header class="header">
        <h2 class="test">Gene Haplotypes</h2>
        <div class="navigation left">
            <a href="./index.html">Home</a>
            <a href="./index.html">About</a>
        </div>
    </header>
    <main class="main">
        <div id="haplotype" class="haplotype-warapper" style="display: block;">
            <div class="haplotype-control">
                <form action="/cgi-bin/hap-show.py" accept-charset="UTF-8" method="post" class="haplotype-form">
                    <input type="text" name="name" class="gene-select" id="name" placeholder="gene name"
                        data-rule="minlen:4" data-msg="Please enter at least 4 chars">
                    <input type="submit" name="commit" value="show">
                </form>
            </div>
            <div class="haplotype-refs" dir="auto">
                <svg width="520" height="40" style="overflow: visible;">
                    <g>
                        <!-- <rect x="131" y="18" width="34" height="10" fill="white" stroke="red"
                            style="stroke-dasharray: 4, 4; stroke-width: 1.5"></rect> -->
                        <!-- 36 -->
                        <circle cx="136" cy="23" r="6" fill="blue"></circle>
                        <rect x="154" y="18" width="34" height="10" fill="red"></rect>
                        <rect x="190" y="18" width="34" height="10" fill="green"></rect>
                        <rect x="226" y="18" width="34" height="10" fill="#800080"></rect>
                        <rect x="262" y="18" width="34" height="10" fill="#FFA500"></rect>
                        <rect x="298" y="18.5" width="34" height="10" fill="#00FFFF"></rect>
                        <rect x="334" y="18" width="34" height="10" fill="#FF4500"></rect>
                        <rect x="370" y="18" width="34" height="10" fill="#1E90FF"></rect>
                        <rect x="406" y="18" width="34" height="10" fill="yellow"></rect>
                    </g>
                    <g transform="translate(130,28)" fill="none" font-size="10" font-family="sans-serif"
                        text-anchor="middle">
                        <!-- 36 -->
                        <g class="tick" opacity="1" transform="translate(5.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">SNP</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(40.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">DEL</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(75.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">INS</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(112.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">SVDEL</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(148.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">SVDUP</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(183.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.70em">UP</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(220.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">UTR</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(256.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">EXON</text>
                        </g>
                        <g class="tick" opacity="1" transform="translate(292.5,0)">
                            <line stroke="currentColor" y2="0"></line>
                            <text fill="currentColor" y="3" dy="0.71em">SVINS</text>
                        </g>
                        <text x="0" y="-16" fill="currentColor" text-anchor="start" font-weight="bold"
                            class="title">Variation And Gene Structure Color Distribution
                        </text>
                    </g>
                </svg>
            </div>
            <div class="haplotype-plot" dir="auto" height="540" style="cursor: pointer; margin: 17px 0;">
                <svg id="d3-plot" class="svgs" height="460" style="background-color: #ffffff;"></svg>
            </div>
            <div class="pagination">
                <button onclick="handleClickToFirstSvg()">&lt;&lt; First</button>
                <button onclick="handleClickToPreSvg()">&lt; Prev</button>
                <span>Page</span>
                <div class="showCurrentPage" contenteditable="true">1</div>
                <span>of</span>
                <span style="margin-right: 4px" class="totalPages"></span>
                <button onclick="handleClickToNextSvg()">Next &gt;</button>
                <button onclick="handleClickToLastSvg()">Last &gt;&gt;</button>
            </div>
            <div id="contentwrap" class="contentwrap">
                <ul class="tab" id="tab">
                    <li id="li_stage" class="li current">
                        <span>INDEL</span>
                    </li>
                    <li id="li_class" class="li"><span>SNP</span></li>
                    <li id="li_photo" class="li"><span>SV</span></li>
                </ul>
                <div class="haplotype-table">
                    <div class="tbl-header">
                        <table>
                            <thead>
                                <tr>
                                    <th>chr</th>
                                    <th>position</th>
                                    <th>length</th>
                                    <th>ref</th>
                                    <th>alt</th>
                                    <th>type</th>
                                    <th>allele</th>
                                    <th>Annotation</th>
                                    <th>Gene_Name</th>
                                    <th>Feature_ID</th>
                                    <th>HGVS.c</th>
                                    <th>HGVS.p</th>
                                </tr>
                            </thead>
                        </table>
                    </div>
                    <div class="tbl-content">
                        <table>
                            <tbody>
                                <tr>
                                    <td>1</td>
                                    <td>LongReach Lancer</td>
                                    <td>40,000,000</td>
                                    <td>485,000,000</td>
                                    <td>445,000,000</td>
                                    <td>1</td>
                                    <td>LongReach Lancer</td>
                                    <td>40,000,000</td>
                                    <td>485,000,000</td>
                                    <td>445,000,000</td>
                                    <td>1</td>
                                    <td>LongReach Lancer</td>
                                </tr>
                                <tr>
                                    <td>1</td>
                                    <td>Jagger</td>
                                    <td>35,000,000</td>
                                    <td>485,000,000</td>
                                    <td>450,000,000</td>
                                    <td>1</td>
                                    <td>LongReach Lancer</td>
                                    <td>40,000,000</td>
                                    <td>485,000,000</td>
                                    <td>445,000,000</td>
                                    <td>1</td>
                                    <td>LongReach Lancer</td>
                                </tr>
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        </div>
    </main>
    <!-- <footer>
        <span>这是footer</span>
    </footer> -->

    <script type="text/javascript">
        let containerId = "d3-plot";
        let container = document.querySelector("#d3-plot");
        let response;
        $.ajax({
            type: "GET",
            //url: "../static/example_update2.json",
            url: "../gene_data/TraesCS2D02G079600.json",
            dataType: "json",
            contentType: "application/json; charset=utf-8",
            async: false,
            success: function (resData) {
                response = resData;
            },
            error: function (err) {
                console.log(err);
            }
        });

        function handleResData(data) {
            let temp = [];
            let geneStructure = data["gene_structure"];
            let sampleNames = data["sample_name"];
            let variation = data["variation"];
            let exon = geneStructure["exon"];
            let upstream = geneStructure["upstream"]["0"];
            let ori = geneStructure["ori"];
            let gene = geneStructure["gene"]["0"];
            let geneName = geneStructure["gene_name"];
            let p5, p3;
            //判断5p_UTR和3p_UTR是否存在
            if (geneStructure["3p_UTR"].length === 0) {
                p3 = [];
            } else {
                p3 = geneStructure["3p_UTR"]["0"];
            }
            if (geneStructure["5p_UTR"].length === 0) {
                p5 = [];
            } else {
                p5 = geneStructure["5p_UTR"]["0"];
            }

            let gs_obj = {};
            if (ori === '+') {
                gs_obj["start"] = upstream["0"];
                gs_obj["end"] = gene["1"];
                gs_obj["geneEnd"] = gene["1"];
            } else if (ori === '-') {
                gs_obj["start"] = gene["0"];
                gs_obj["end"] = upstream["1"];
                gs_obj["geneEnd"] = gene["1"];
            }

            if (+(gs_obj["start"]) > 0) {
                if (p3.length != 0) {
                    gs_obj["3p_UTR"] = p3["1"] - p3["0"] + 1 + gs_obj["start"];
                } else {
                    gs_obj["3p_UTR"] = 0;
                }
                if (p5.length != 0) {
                    gs_obj["5p_UTR"] = p5["1"] - p5["0"] + 1 + gs_obj["start"];
                } else {
                    gs_obj["5p_UTR"] = 0;
                }
                gs_obj["upstream"] = upstream["1"] - upstream["0"] + 1 + gs_obj["start"];
                let _exon = [];
                exon.forEach(d => {
                    _exon.push([d["0"], d["1"] + gs_obj["start"]]);

                })
                gs_obj["exon"] = _exon;
            } else {
                if (p3.length != 0) {
                    gs_obj["3p_UTR"] = p3["1"] - p3["0"] + 1
                } else {
                    gs_obj["3p_UTR"] = 0;
                }
                if (p5.length != 0) {
                    gs_obj["5p_UTR"] = p5["1"] - p5["0"] + 1;
                } else {
                    gs_obj["5p_UTR"] = 0;
                }
                gs_obj["upstream"] = upstream["1"] - upstream["0"] + 1;
                gs_obj["exon"] = exon;
            }
            gs_obj["name"] = sampleNames;
            gs_obj["variation"] = variation;
            gs_obj["ori"] = ori;
            gs_obj["geneName"] = geneName;
            temp.push(gs_obj);
            return temp;
        }

        let data = handleResData(response);
        console.log(data);

        // 常量
        let margin = ({
            top: 50,
            right: 100,
            bottom: 10,
            left: 60
        })
        let color = ({
            ins: 'green',
            del: 'red',
            utrP5: '#FF4500',
            utrP3: '#FF4500',
            upstream: '#00FFFF',
            snp: 'blue',
            exon: '#1E90FF',
            svDup: '#FFA500',
            svDel: '#800080',
            svIns: 'yellow',
        })
        let svg = d3.select(container);
        // width="1450" height="360" 
        // let width = +svg.attr("width");
        let height = +svg.attr("height");
        let width = $(".haplotype-plot").width();
        // let height = data[0].name.length * 65 / 50;
        // svg.attr("height", height);
        svg.attr("width", width);

        $(window).resize(function () {
            width = $(window).width();
            svg.attr("width", width);
            location.reload();
        })
        let innerWidth = width - margin.left - margin.right;
        let innerHeight = height - margin.top - margin.bottom;
        let xScale, yScale;
        let lineLeft, lineRight, geneLength;
        let xScaleMax, start, ori;
        let gsHeight = 35;
        let gsHeightMin = 15;
        let snp_r = 3;
        let pathMoveY = 10;
        let delta = 50;
        let snpPosition = [];
        let _snp, graph, bindHeight, sampleNames, geneEnd, geneName;
        let yValue = (datum) => { return datum.name };

        ori = data[0].ori;
        start = data[0].start;
        geneEnd = data[0].geneEnd;
        geneName = data[0].geneName;
        if (ori === '+') {
            xScaleMax = data[0].end;
        } else if (ori === '-') {
            xScaleMax = data[0].end;
        }

        const renderInit = function (data) {
            // sampleNames is dynamic
            sampleNames = data[0].name;
            //设置坐标轴的比例尺
            xScale = d3.scaleLinear()
                //.domain([d3.min(data, xValue), d3.max(data, xValue)])
                .domain([start, xScaleMax])
                .range([0, innerWidth]);

            yScale = d3.scaleBand()
                // .domain(d3.map(yValue(data[0]), d => d))
                .domain(sampleNames)
                .range([0, innerHeight]);

            const g = svg.append("g")
                .attr("transform", `translate(${margin.left}, ${margin.top})`)
                .attr("id", "maingroup")

            //绘制坐标轴
            const xAxis = d3.axisTop(xScale);
            let xAxisGroup = g.append('g').call(xAxis).attr('transform', `translate(${0}, ${0})`).attr("class", "top-axis")
                .attr("text-anchor", "end");
            const yAxis = d3.axisLeft(yScale);
            g.append('g').attr("class", "sample-name").call(yAxis).attr("class", "left-axis");

            //设置Variation
            variation = data[0].variation;
            let xAxisGroupTop = g.append('g').call(xAxis).attr('transform', `translate(${0}, ${-30})`)
                .attr("text-anchor", "end");

            //绘制版心
            graph = g.append("g").attr("class", "map")
            bindHeight = yScale.bandwidth();

            //所有的基因柱状底色全为灰色
            let bottomColor = [];

            for (let i = 0; i < sampleNames.length; i++) {
                let bind = [start, sampleNames[i], gsHeightMin, xScaleMax]
                bottomColor.push(bind);
            }

            bottomColor.forEach(d => {
                graph.append("rect").attr("fill", 'gray').attr("class", "bc")
                    .attr("opacity", 0.6).attr("x", xScale(d[0]))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3]));
            })

            //绘制不同基因区间的图形
            //绘制upstream
            upstream = [];
            if (ori === '+') {
                for (let i = 0; i < sampleNames.length; i++) {
                    let bind = [start, sampleNames[i], gsHeightMin, data[0].upstream]
                    upstream.push(bind);
                }
            } else if (ori === '-') {
                for (let i = 0; i < sampleNames.length; i++) {
                    let bind = [geneEnd, sampleNames[i], gsHeightMin, data[0].upstream]
                    upstream.push(bind);
                }
            }

            upstream.forEach(d => {
                graph.append("rect").attr("fill", color.upstream).attr("opacity", 1).attr("x", xScale(d[0]))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3]))
            })

            //绘制5p_UTR
            utrP5 = [];
            if (data[0]["5p_UTR"] != 0) {
                if (ori === '+') {
                    for (let i = 0; i < sampleNames.length; i++) {
                        let bind = [data[0].upstream + start, sampleNames[i], gsHeight, data[0]["5p_UTR"]]
                        utrP5.push(bind);
                    }
                } else if (ori === '-') {
                    for (let i = 0; i < sampleNames.length; i++) {
                        let bind = [start, sampleNames[i], gsHeight, data[0]["5p_UTR"]]
                        utrP5.push(bind);
                    }
                }

                utrP5.forEach(d => {
                    graph.append("rect").attr("fill", color.utrP5).attr("opacity", 1).attr("x", xScale(d[0]))
                        .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2)
                        .attr("height", d[2]).attr("width", xScale(d[3]))
                })
            }

            //绘制3p_UTR
            utrP3 = [];
            for (let i = 0; i < sampleNames.length; i++) {
                let bind = [geneEnd - data[0]["3p_UTR"] + start, sampleNames[i], gsHeight, data[0]["3p_UTR"]]
                utrP3.push(bind);
            }

            utrP3.forEach(d => {
                graph.append("rect").attr("fill", color.utrP3).attr("opacity", 1).attr("x", xScale(d[0]))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3]))
            })

            //绘制exon 
            exon = [];
            for (let i = 0; i < sampleNames.length; i++) {
                let exonData = data[0].exon;
                exonData.forEach(d => {
                    let innerBind = [d[0], sampleNames[i], gsHeight, d[1] - d[0] + 1];
                    exon.push(innerBind);
                })
            }
            exon.forEach(d => {
                graph.append("rect").attr("fill", color.exon).attr("opacity", 1).attr("x", xScale(d[0]))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3]))
            })

            //绘制INDEL插入缺失
            del = [];
            ins = [];
            INDEL = variation.INDEL;
            for (let i = 0; i < sampleNames.length; i++) {
                for (let j = 0; j < INDEL.length; j++) {
                    let item = INDEL[j];
                    if (item.type == 'DEL') {
                        if (item.stat_in_each_sample[i] == '1') {
                            let bind = [item.position, sampleNames[i], gsHeightMin, item.length];
                            del.push(bind);
                        }
                    } else if (item.type == 'INS') {
                        if (item.stat_in_each_sample[i] == '1') {
                            let bind = [item.position, sampleNames[i], gsHeightMin, item.length];
                            ins.push(bind);
                        }
                    }
                }
            }

            del.forEach(d => {
                graph.append("rect").attr("stroke", color.del).attr("fill", "white").attr("opacity", 1).attr("x", xScale(d[0] - d[3] / 2))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3] + start))
                    .style("stroke-dasharray", "4, 4").style("stroke-width", '1.5');

                // top graph
                xAxisGroupTop.append('rect').attr("x", xScale(d[0] - d[3] / 2)).attr("y", "-4").attr("stroke", color.del).attr("fill", "white")
                    .attr("width", xScale(d[3] + start)).attr("height", "8").attr("text-anchor", "middle")
                    .style("stroke-dasharray", "4, 4").style("stroke-width", '1.5');
            })

            ins.forEach(d => {
                graph.append("path").attr("class", "ins").attr("fill", "none").attr("stroke", color.ins).attr("stroke-width", 2)
                    .attr("d", `M ${xScale(d[0] - d[3] / 2)} ${yScale(d[1]) + d[2] * -0.1} L ${xScale(d[0] + d[3] / 2)} ${yScale(d[1]) + d[2] * -0.1} 
                        L ${xScale(d[0])} ${yScale(d[1]) + d[2] * -0.1} L ${xScale(d[0])} ${yScale(d[1]) + bindHeight / 2}`)

                xAxisGroupTop.append("path").attr("class", "ins").attr("fill", "none").attr("stroke", color.ins).attr("stroke-width", 2)
                    .attr("d", `M ${xScale(d[0] - delta)} 7 L ${xScale(d[0] + delta)} 7
                        L ${xScale(d[0])} -10 L ${xScale(d[0] - delta)} 7`).attr("text-anchor", "middle");
            })

            //绘制sv
            svDel = [];
            svDup = [];
            svIns = [];
            SV = variation.SV;
            for (let i = 0; i < sampleNames.length; i++) {
                for (let j = 0; j < SV.length; j++) {
                    let item = SV[j];
                    if (item.type == 'DEL') {
                        if (item.stat_in_each_sample[i] == '1') {
                            let bind = [item.position, sampleNames[i], gsHeightMin, item.length];
                            svDel.push(bind);
                        }
                    } else if (item.type == 'DUP') {
                        if (item.stat_in_each_sample[i] == '1') {
                            let bind = [item.position, sampleNames[i], gsHeightMin, item.length];
                            svDup.push(bind);
                        }
                    } else if (item.type == 'INS') {
                        if (item.stat_in_each_sample[i] == '1') {
                            let bind = [item.position, sampleNames[i], gsHeightMin, item.length];
                            svIns.push(bind);
                        }
                    }
                }
            }

            svIns.forEach(d => {
                graph.append("path").attr("class", "svins").attr("fill", "none").attr("stroke", color.svIns).attr("stroke-width", 2)
                    .attr("d", `M ${xScale(d[0])} ${yScale(d[1]) + pathMoveY} L ${xScale(d[0] + d[3])} ${yScale(d[1]) + pathMoveY} 
                        L ${xScale(d[0] + d[3] / 2)} ${yScale(d[1]) + pathMoveY} L ${xScale(d[0] + d[3] / 2)} ${yScale(d[1]) + bindHeight / 2 - d[2] / 2}`)

                xAxisGroupTop.append("path").attr("class", "svins").attr("fill", "none").attr("stroke", color.svIns).attr("stroke-width", 2)
                    .attr("d", `M ${xScale(d[0] - delta)} 7 L ${xScale(d[0] + delta)} 7
                        L ${xScale(d[0])} -10 L ${xScale(d[0] - delta)} 7`).attr("text-anchor", "middle");
            })

            svDel.forEach(d => {
                graph.append("rect").attr("stroke", color.svDel).attr("fill", "white").attr("opacity", 1).attr("x", xScale(d[0] - d[3] / 2))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3] + start))
                    .style("stroke-dasharray", "4, 4").style("stroke-width", '1.5');

                xAxisGroupTop.append('rect').attr("x", xScale(d[0] - d[3] / 2)).attr("y", "-2").attr("stroke", color.svDel).attr("fill", "white")
                    .attr("width", xScale(d[3] + start)).attr("height", "6").attr("text-anchor", "middle")
                    .style("stroke-dasharray", "4, 4").style("stroke-width", '1.5');
            })
            svDup.forEach(d => {
                graph.append("rect").attr("fill", color.svDup).attr("opacity", 1).attr("x", xScale(d[0] - d[3] / 2))
                    .attr("y", yScale(d[1]) + bindHeight / 2 - d[2] / 2).attr("height", d[2]).attr("width", xScale(d[3] + start));
                xAxisGroupTop.append('rect').attr("x", xScale(d[0] - d[3] / 2)).attr("y", "-4").attr("fill", color.svDup)
                    .attr("width", xScale(d[3] + start)).attr("height", "8").attr("text-anchor", "middle");
            })

            //绘制CN 
            _cn = [];
            CN = variation.CN;
            for (let i = 0; i < sampleNames.length; i++) {
                let text = [xScaleMax + 150, sampleNames[i], gsHeightMin, CN[i]];
                _cn.push(text);
            }

            _cn.forEach(d => {
                graph.append("text").attr("opacity", 1).attr("x", xScale(d[0]))
                    .attr("y", yScale(d[1]) + bindHeight / 2 + 10).text(d[3])
                    .attr("font-size", '1.5em').attr("text-anchor", 'middle');
            })

            //draw SNP
            _snp = [];
            SNP = variation.SNP;
            for (let i = 0; i < sampleNames.length; i++) {
                for (let j = 0; j < SNP.length; j++) {
                    let item = SNP[j];
                    let bind = [item.position, sampleNames[i], gsHeightMin, item.stat_in_each_sample[i]];
                    _snp.push(bind);
                    snpPosition.push(item.position);
                }
            }
            _snp.forEach(d => {
                graph.append("text").attr("opacity", 0).attr("x", xScale(d[0])).attr("class", "snp")
                    .attr("y", yScale(d[1]) + bindHeight / 2 + 10).text(d[3]).attr("class", "snp")
                    .attr("font-size", '1em').attr("text-anchor", 'start');
                xAxisGroupTop.append('circle').attr("cx", xScale(d[0])).attr("cy", "0").attr("r", snp_r).attr("fill", color.snp);
            })
        }

        let createTooltip = function (el) {
            el
                .style("position", "absolute")
                .style("pointer-events", "none")
                .style("top", 0)
                .style("opacity", 0)
                .style("background", "red")
                .style("border-radius", "5px")
                .style("box-shadow", "0 0 10px rgba(0,0,0,.25)")
                .style("padding", "10px")
                .style("line-height", "1.3")
                .style("font", "11px sans-serif")
        }

        // show text following line 
        const showText = function (xPosition, yPosition, withMouseOut) {
            let textShowDeviation = 120;
            let xPositionMappingAxis = Math.round(xPosition * (xScaleMax - start) / geneLength) + start;
            if (withMouseOut !== undefined) {
                if (withMouseOut) {
                    svg.select("text.show-text")
                        .text("")
                        .attr("x", xPosition)
                        .attr("y", yPosition)
                }
            } else {
                if (xPositionMappingAxis > xScaleMax / 2) {
                    svg.select("text.show-text")
                        .text(xPositionMappingAxis + 3)
                        .attr("x", xPosition + textShowDeviation)
                        .attr("y", margin.top - 5)
                } else {
                    svg.select("text.show-text")
                        .text(xPositionMappingAxis + 1)
                        .attr("x", xPosition + textShowDeviation)
                        .attr("y", margin.top - 5)
                }

            }
        }
        // array Deduplication
        function unique(arr) {
            return Array.from(new Set(arr));
        }
        // When the SNP mutation site is detected, the mutation information is displayed 
        const showSNPMutationInfo = function (xPosition) {
            let changeToXScale = Math.round((xPosition - lineLeft));

            let allSNP = [];
            $(".snp").each(function () {
                allSNP.push($(this).attr("x"));
            });
            let uniqueSNP = unique(allSNP);
            let changeSNP = uniqueSNP.map(item => Math.round(item));

            // snpPosition.includes(changeToXScale)
            changeSNP.forEach(item => {
                if (changeToXScale - snp_r * 2 < item && item < changeToXScale + snp_r * 2) {
                    ix = changeSNP.indexOf(item);
                    currentX = uniqueSNP[ix];
                    $(".snp").each(function () {
                        if ($(this).attr("x") === currentX) {
                            $(this).attr("opacity", 1)
                        } else (
                            $(this).attr("opacity", 0)
                        )
                    });
                } else {
                    $(".snp").each(function () {
                        $(this).attr("opacity", 0)
                    });
                }
            })
        }

        const drawLine = function () {
            lineLeft = margin.left;
            geneLength = +($(".bc").attr("width"));
            lineRight = geneLength + lineLeft;
            const tooltip = d3.select(document.createElement("div")).call(createTooltip);
            const hoverLine = svg.append("line").attr("class", "hover-line").attr("y1", 0)
                .attr("y2", height - margin.bottom).attr("stroke", "red").style("pointer-events", "none").style("opacity", 0);

            const hoverText = svg.append("svg:g").attr("class", "hover-text").append("svg:text").attr("class", "show-text")
                .attr("font-size", 16).attr("fill", "#333").text("").attr("x", 0).attr("y", 0);
            svg
                .on("mousemove", function (event) {
                    // get current mouse location, return [mouseX, mouseY]
                    let [x, y] = d3.pointer(event);

                    // draw line
                    if (x < lineLeft) {
                        hoverLine.style("opacity", 0);
                        hoverText.style("opacity", 0);
                    } else if (x >= lineLeft && x <= lineRight) {
                        hoverLine.style("opacity", 1);
                        hoverLine.attr("transform", `translate(${x}, 0)`);
                        hoverText.style("opacity", 1);
                        showText(x - lineLeft, y);
                        showSNPMutationInfo(x);
                    } else if (x > lineRight) {
                        hoverLine.style("opacity", 0);
                        hoverText.style("opacity", 0);
                    }

                    if (y < margin.top | y > height - margin.bottom) {
                        hoverText.style("opacity", 0);
                        showText(0, 0, true);
                    }

                    y += 20;
                    if (x > width / 2) x -= 100;

                    tooltip
                        .style("left", x + "px")
                        .style("top", y + "px")

                })
        }

        let pageMaxNumber = 10;
        let currentPageNumber;
        let pages = (data[0].name.length % pageMaxNumber) === 0 ? (data[0].name.length / pageMaxNumber) :
            Math.floor(data[0].name.length / pageMaxNumber) + 1;

        $(".totalPages").text(pages);

        // 处理variation的数据
        function handleDataCutOut(arr, start, end) {
            /*
            arr: 处理的数组; start: 截取开始的位置; end: 截取结束的位置
            */
            let rt = [];
            arr.forEach(d => {
                let svTemp = {};
                for (key in d) {
                    if (key === "stat_in_each_sample") {
                        svTemp[key] = d[key].slice(start, end);
                    } else {
                        svTemp[key] = d[key];
                    }
                }
                rt.push(svTemp);
            })
            return rt;
        }

        function firstPageDataInitAndDataPagination(data, start, end) {
            // 返回array，返回的数据包括：
            //sampleNames:[], exon:[[]], variation: {CN:[],INDEL:[{}],SNP:[{}],SV:[{}]}, 3p, 5p, upstream
            if (start === undefined) {
                start = 0;
            }
            if (end === undefined) {
                end = pageMaxNumber;
            }
            let firstPageData = [];
            let fpGs = {};
            let _variation = {};
            fpVariation = data[0].variation;
            fpGs["name"] = data[0].name.slice(start, end);
            fpGs["exon"] = data[0].exon;
            fpGs["3p_UTR"] = data[0]["3p_UTR"];
            fpGs["5p_UTR"] = data[0]["5p_UTR"];
            fpGs["upstream"] = data[0].upstream;
            _variation["CN"] = fpVariation.CN.slice(start, end);
            _variation["SNP"] = handleDataCutOut(fpVariation.SNP, start, end);
            _variation["SV"] = handleDataCutOut(fpVariation.SV, start, end);
            _variation["INDEL"] = handleDataCutOut(fpVariation.INDEL, start, end);
            fpGs["variation"] = _variation;
            firstPageData.push(fpGs);
            return firstPageData;
        }

        function removeExistElements() {
            d3.select("#maingroup").remove();
            d3.select(".hover-line").remove();
            d3.select(".hover-text").remove();
        }

        function getCurrentPageNumber() {
            return +($(".showCurrentPage").text());
        }

        function setCurrentPageNumber(num) {
            $(".showCurrentPage").text(num);
        }

        if (pages === 1) {
            let pageData = firstPageDataInitAndDataPagination(data);
            renderInit(pageData);
            drawLine();
            setCurrentPageNumber(1);
        } else {
            // 对样本数超过10的gene结构进行绘制svg
            currentPageNumber = getCurrentPageNumber();
            let pageData = firstPageDataInitAndDataPagination(data, 0, pageMaxNumber);
            renderInit(pageData);
            drawLine();
            setCurrentPageNumber(currentPageNumber);
        }

        function handleClickToFirstSvg() {
            // show first svg page
            removeExistElements();
            let pageData = firstPageDataInitAndDataPagination(data, 0, pageMaxNumber);
            renderInit(pageData);
            drawLine();
            //location.reload();
            setCurrentPageNumber(1);
        }
        function handleClickToPreSvg() {
            // pre page
            currentPageNumber = getCurrentPageNumber();
            if (currentPageNumber != 1) {
                removeExistElements();
                let pageData = firstPageDataInitAndDataPagination(data,
                    (currentPageNumber - 2) * pageMaxNumber, (currentPageNumber - 1) * pageMaxNumber);
                renderInit(pageData);
                drawLine();
                //location.reload();
                setCurrentPageNumber(currentPageNumber - 1);
            }
        }
        function handleClickToNextSvg() {
            // next page
            currentPageNumber = getCurrentPageNumber();
            if (currentPageNumber != pages) {
                removeExistElements();
                let pageData = firstPageDataInitAndDataPagination(data,
                    currentPageNumber * pageMaxNumber, (currentPageNumber + 1) * pageMaxNumber);
                renderInit(pageData);
                drawLine();
                //location.reload();
                setCurrentPageNumber(currentPageNumber + 1);
            }
        }
        function handleClickToLastSvg() {
            removeExistElements();
            let pageData = firstPageDataInitAndDataPagination(data, (pages - 1) * pageMaxNumber, pages * pageMaxNumber);
            renderInit(pageData);
            drawLine();
            //location.reload();
            setCurrentPageNumber(pages);
        }
        function handleClickToInputSvg(page) {
            removeExistElements();
            let pageData = firstPageDataInitAndDataPagination(data,
                (page - 1) * pageMaxNumber, (page) * pageMaxNumber);
            renderInit(pageData);
            drawLine();
            //location.reload();
            setCurrentPageNumber(page);
        }
        $(".showCurrentPage").keyup(function (event) {
            if (event.keyCode == 13) {
                currentPageNumber = getCurrentPageNumber();
                handleClickToInputSvg(currentPageNumber);
            }
        });

        // show table 
        $('#tab > li').on("click", function () {
            var num = $(this).index();
            //var src=basePath+urlArr[num];
            $('#tab > li').each(function () {
                if ($(this).index() == num) {
                    $(this).attr("class", "li current");
                    if (num == 1) {
                        csvDataToHtml("snp");
                    } else if (num == 2) {
                        csvDataToHtml("sv");
                    } else {
                        csvDataToHtml("indel");
                    }
                } else {
                    $(this).attr("class", "li");
                }
            });
        });
        function showTableInit() {

        }
        function csvDataToHtml(tab) {
            console.log(tab);
            let csvData;
            //let csvPath = "../gene_data/" + geneName.slice(0, -1) + tab + ".csv";
            let csvPath = "../gene_data/TraesCS2D02G079600.snp.csv"
            if (tab === "snp") {
                csvData = d3.csv(csvPath);
                console.log(csvData)
            } else if (tab === "sv") {

            } else if (tab === "indel") {

            }
        }
    </script>
</body>

</html>
